src/encoding.rs

```diff
@@ -815,11 +815,23 @@ pub mod string {
     where
         B: Buf,
     {
-        let mut value_bytes = mem::replace(value, String::new()).into_bytes();
-        bytes::merge(wire_type, &mut value_bytes, buf, ctx)?;
-        *value = String::from_utf8(value_bytes)
-            .map_err(|_| DecodeError::new("invalid string value: data is not UTF-8 encoded"))?;
-        Ok(())
+        unsafe {
+            // String::as_mut_vec is unsafe because it doesn't check that the bytes
+            // inserted into it the resulting vec are valid UTF-8. We check
+            // explicitly in order to ensure this is safe.
+            let value = value.as_mut_vec();
+            super::bytes::merge(wire_type, value, buf, ctx)?;
```

Comment:
This still leaves the `String` value in invalid state if any method on `buf` panics.
A scope guard, like originally proposed in 55f644e7401c3cf0dfef2b078a73c5362641f6a2 for #194 but working on the `&mut Vec<u8>` to avoid technical unsoundness, would be more bulletproof.